dgidb/graph_app.py

```diff
@@ -46,26 +68,107 @@ def set_app_layout(app,plot,genes):
+            return ng.generate_json(updated_graph), updated_plot
+        return None, ng.generate_plotly(None)
+    
+def update_selected_node(app):
+    @app.callback(
+        Output('selected-node', 'data'),
+        Input('network-graph', 'clickData')
+    )
+    def update(clickData):
+        if clickData is not None and 'points' in clickData:
+            selected_node = clickData['points'][0]
+            return selected_node
+        return dash.no_update
+
+def update_selected_node_display(app):
+    @app.callback(
+        Output('selected-node-text', 'children'),
+        Input('selected-node', 'data')
+    )
+    def update(selected_node):
+        return selected_node['text']
+        print(selected_node)
+
+def update_neighbor_dropdown(app):
+    @app.callback(
+        Output('neighbor-dropdown', 'options'),
+        Input('selected-node', 'data')
+    )
+    def update(selected_node):
+        if(selected_node['curveNumber'] != 1):
+            return selected_node['customdata']
+        else:
+            return []
+
+def update_edge_info(app):
+    @app.callback(
+        Output('edge-info-text', 'children'),
+        [Input('selected-node', 'data'), Input('neighbor-dropdown', 'value')],
+        State('graph', 'data')
+    )
+    def update(selected_node,selected_neighbor,graph):
+        if selected_node is None:
+            return dash.no_update
+        if(selected_node['curveNumber'] == 1):
+            selected_data = get_node_data_from_id(graph['links'], selected_node['text'])
+            return "ID: " + str(selected_data['id']) + "\n\nApproval: " + str(selected_data['approval']) + "\n\nScore: " + str(selected_data['score']) + "\n\nAttributes: " + str(selected_data['attributes']) + "\n\nSource: " + str(selected_data['source']) + "\n\nPmid: " + str(selected_data['pmid'])
+        if selected_neighbor is not None:
+            edge_node_id = None
+            selected_node_is_gene = get_node_data_from_id(graph['nodes'], selected_node['text'])['isGene']
+            selected_neighbor_is_gene = get_node_data_from_id(graph['nodes'], selected_neighbor)['isGene']
+            if(selected_node_is_gene == selected_neighbor_is_gene):
+                return dash.no_update
+            elif(selected_node_is_gene):
+                edge_node_id = selected_node['text'] + " - " + selected_neighbor
+            elif(selected_neighbor_is_gene):
+                edge_node_id = selected_neighbor + " - " + selected_node['text']
+
+            selected_data = get_node_data_from_id(graph['links'], edge_node_id)
+            if selected_data is None:
+                return dash.no_update
+            return "ID: " + str(selected_data['id']) + "\n\nApproval: " + str(selected_data['approval']) + "\n\nScore: " + str(selected_data['score']) + "\n\nAttributes: " + str(selected_data['attributes']) + "\n\nSource: " + str(selected_data['source']) + "\n\nPmid: " + str(selected_data['pmid'])
+        return dash.no_update

```

Comment:
I think the error posted in the comments happens because of something not properly firing within this block under certain conditions.